### PR TITLE
Expand testing

### PR DESCRIPTION
**Pull request by Joao-Dionisio**
Some notes and questions.

There is no method to check whether a solution has been freed. If we try to ```setSolVal``` in a freed solution, we get a segfault. 

```addConsLocal``` appears to still be causing issues.

```addConsIndicator``` appears to still be causing issues when the binary variable is $0$. Should we add tests that we know will fail because the methods are wrong?

I think logical constraints were already tested in ```test_logical.py```, but the syntax is completely different from the rest of the tests. 

Should we treat the tests as a potential tutorial? I know I used ```test_pricer.py``` to know how a pricer works, but I've now added some methods there (```getNSolsFound```, ```getNBestSolsFound```, ```getNSols```) just to test them, and I worry that new users would find it weird. In my opinion, we should greatly expand the tutorial folder, and assume that the tests will not be accessed by regular users.

### CODE REVIEW

File: tests/test_cons.py [hunk 1]
```diff
@@ -39,4 +39,94 @@ def test_constraint_option_setting():
...
+    m.addConsXor([x1,x3], True)
+
+    # result2 false
+    m.addConsOr([x3,x4], result2)
+    m.addConsAnd([x1,x3], result2)
+    m.addConsXor([x1,x2], False)
+
+    m.optimize() 
+
+    assert m.getVal(result1) == 1
```
Comment (mmghannam) [inline tests/test_cons.py:68]: Should be compared with `isEQ`

File: tests/test_cons.py [hunk 2]
```diff
@@ -39,4 +39,94 @@ def test_constraint_option_setting():
...
+
+    m.addVarSOS1(c1, x[2], 1)
+    m.addVarSOS2(c2, x[3], 1)
+
+    m.appendVarSOS1(c1, x[4])
+    m.appendVarSOS2(c2, x[5])
+
+    m.optimize()
+
+    assert m.getVal(x[0]) == 0
```
Comment (mmghannam) [inline tests/test_cons.py:88]: same here

File: tests/test_cons.py [hunk 3]
```diff
@@ -39,4 +39,94 @@ def test_constraint_option_setting():
...
+    assert m.getVal(x[1]) == 0
+    assert m.getVal(x[2]) == 0
+    assert m.getVal(x[3]) == 1
+    assert m.getVal(x[4]) == 1
+    assert m.getVal(x[5]) == 1
+
+
+# WIP. addConsIndicator doesn't behave as expected when binary variable is False. See Issue #717.
```
Comment (mmghannam) [inline tests/test_cons.py:96]: Maybe change it to an xfail test.

File: tests/test_solution.py [hunk 4]
```diff
@@ -32,9 +32,19 @@ def test_solution_create():
 
     s = m.createSol()
     s[x] = 2.0
-    s[y] = 4.0
+    s[y] = 5.0
+    assert not m.checkSol(s)
     assert m.addSol(s, free=True)
 
+    s1 = m.createSol()
+    m.setSolVal(s1,x,1.0)
```
Comment (mmghannam) [inline tests/test_solution.py:40]: Missing spaces, use an auto-formatting tool 😄

File: tests/test_solution.py [hunk 5]
```diff
@@ -32,9 +32,19 @@ def test_solution_create():
...
+    m.setSolVal(s1,y,2.0)
+    assert m.checkSol(s1)
+
+    m.optimize()
+
+    assert m.getSolObjVal(s1) == -1
```
Comment (mmghannam) [inline tests/test_solution.py:46]: With `isEQ` please

File: tests/test_solution.py [hunk 6]
```diff
@@ -59,9 +69,13 @@ def test_solution_evaluation():
 
     # Check consistency with Models's getVal method
     assert sol[x] == m.getVal(x)
+    assert m.getSolVal(sol,x) == m.getVal(x)
```
Comment (mmghannam) [inline tests/test_solution.py:72]: here too please and also formatting.

File: tests/test_cons.py (hunk 1 above)
Comment (Joao-Dionisio) [reply tests/test_cons.py]: Oh yeah, thanks Mo!

File: tests/test_cons.py (hunk 3 above)
Comment (Joao-Dionisio) [reply tests/test_cons.py]: Didn't know about this, thanks for letting me know!

File: tests/test_cons.py [hunk 7]
```diff
@@ -39,4 +42,120 @@ def test_constraint_option_setting():
...
+
+    m.optimize()
+
+    assert m.isEQ(m.getVal(slack), 0)
+    assert m.isEQ(m.getVal(binvar), 1)
+    assert m.isEQ(m.getVal(x), 1)
+
+
+# addConsIndicator doesn't behave as expected when binary variable is False. See Issue #717.
+@pytest.mark.xfail
```
Comment (mmghannam) [inline tests/test_cons.py:118]:
```suggestion
@pytest.mark.xfail(reason="addConsIndicator doesn't behave as expected when binary variable is False. See Issue #717."
```

File: tests/test_cons.py [hunk 8]
```diff
@@ -39,4 +42,120 @@ def test_constraint_option_setting():
...
+    slack = m.getSlackVarIndicator(c)
+
+    m.optimize()
+
+    assert m.isEQ(m.getVal(slack), 0)
+    assert m.isEQ(m.getVal(binvar), 1)
+    assert m.isEQ(m.getVal(x), 1)
+
+
+# addConsIndicator doesn't behave as expected when binary variable is False. See Issue #717.
```
Comment (mmghannam) [inline tests/test_cons.py:117]:
```suggestion
```

File: tests/test_cons.py [hunk 9]
```diff
@@ -39,4 +42,119 @@ def test_constraint_option_setting():
...
+    slack = m.getSlackVarIndicator(c)
+
+    m.optimize()
+
+    assert m.isEQ(m.getVal(slack), 0)
+    assert m.isEQ(m.getVal(binvar), 1)
+    assert m.isEQ(m.getVal(x), 1)
+
+
+@pytest.mark.xfail(reason="addConsIndicator doesn't behave as expected when binary variable is False. See Issue #717."
```
Comment (mmghannam) [inline tests/test_cons.py:117]:
```suggestion
@pytest.mark.xfail(reason="addConsIndicator doesn't behave as expected when binary variable is False. See Issue #717.")
```

File: tests/test_cons.py [hunk 10]
```diff
@@ -39,4 +42,119 @@ def test_constraint_option_setting():
...
+def test_printCons():
+    m = Model()
+    x = m.addVar()
+    y = m.addVar()
+    c = m.addCons(x * y <= 5)
+
+    m.printCons(c)
+
+
+@pytest.mark.skip
```
Comment (mmghannam) [inline tests/test_cons.py:153]:
```suggestion
@pytest.mark.skip(reason="TODO: test getValsLinear()")
```

File: tests/test_cons.py [hunk 11]
```diff
@@ -39,4 +42,119 @@ def test_constraint_option_setting():
...
+def test_getValsLinear():
+    assert True
+
+
+@pytest.mark.skip
```
Comment (mmghannam) [inline tests/test_cons.py:158]:
```suggestion
@pytest.mark.skip(reaons="TODO: test getRowLinear()")
```

**Review by mmghannam — APPROVED** [review]

**Pull request merged**